include/ginkgo/core/log/batch_logger.hpp

```diff
@@ -0,0 +1,163 @@
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_LOG_BATCH_LOGGER_HPP_
+#define GKO_PUBLIC_CORE_LOG_BATCH_LOGGER_HPP_
+
+
+#include <memory>
+
+
+#include <ginkgo/core/base/batch_multi_vector.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/log/logger.hpp>
+
+
+namespace gko {
+namespace batch {
+/**
+ * @brief The logger namespace .
+ * @ref log
+ * @ingroup log
+ */
+namespace log {
+
+
+/**
+ * Types of batch loggers available.
+ */
+enum class BatchLogType {
+    convergence_completion,
+    simple_convergence_completion
+};
+
+
+struct BatchLogDataBase {
+    /**
+     * Stores convergence iteration counts for every matrix in the batch and
+     * for every right-hand side.
+     */
+    array<int> iter_counts;
+
+    virtual ~BatchLogDataBase() = default;
+};
+
+
+/**
+ * Stores logging data for batch solver kernels.
+ */
+template <typename ValueType>
+struct BatchLogData : public BatchLogDataBase {
+    /**
+     * Stores residual norm values for every linear system in the batch
+     * for every right-hand side.
+     */
+    std::shared_ptr<MultiVector<remove_complex<ValueType>>> res_norms;
+};
+
+
+/**
+ * Logs the final residuals and iteration counts for a batch solver.
+ *
+ * The purpose of this logger is to give simple access to standard data
+ * generated by the solver once it has converged.
+ *
+ * @note This logger copies the data (iterations and residual norm) to the host
+ * executor.
+ *
+ * @ingroup log
+ */
+template <typename ValueType = default_precision>
+class BatchConvergence : public gko::log::Logger {
+public:
+    using real_type = remove_complex<ValueType>;
+    using mask_type = gko::log::Logger::mask_type;
+
+    void on_batch_solver_completed(
+        const array<int>& iteration_count,
+        const MultiVector<real_type>* residual_norm) const override;
+
+    /**
+     * Creates a convergence logger. This dynamically allocates the memory,
+     * constructs the object and returns an std::unique_ptr to this object.
+     *
+     * @param exec  the executor
+     * @param enabled_events  the events enabled for this logger. By default all
+     *                        events.
+     *
+     * @return an std::unique_ptr to the the constructed object
+     */
+    static std::unique_ptr<BatchConvergence> create(
+        std::shared_ptr<const Executor> exec,
```

Comment:
That is true. But I think we have to remember that this means that the `on_<>` event will need to deep copy the data, with possible allocation. This might significantly affect performance. It might be better to have the executor as a parameter and even allocate the internal objects if their sizes are known.

Reply:
I mean by default when the Logger is created, the arrays are null. The internal objects are created only when the `event` is called. So there will have to be allocations.

Reply:
Yes you would have only one allocation in both cases, assuming that the number of batches and the host executor doesn't change.